redsky/event_streams.py

```diff
@@ -0,0 +1,633 @@
+    def wraps(kwargs1, **kwargs2):
+        kwargs1.update(kwargs2)
+        return f(**kwargs1)
+
+    return wraps
+
+
+class EventStream(Stream):
+    def __init__(self, child=None, children=None,
+                 output_info=None, input_info=None,
+                 **kwargs):
+        """
+        Serve up documents and their internals as requested.
+        The main way that this works is by a) ingesting documents, b) issuing
+        documents, c) returning the internals of documents upon request.
+
+        Parameters
+        ----------
+        input_info: list of tuples
+            describs the incoming streams
+        output_info: list of tuples
+            describs the resulting stream
+        provenance : dict, optional
+            metadata about this operation
+
+        Notes
+        ------
+        input_info is designed to map keys in streams to kwargs in functions.
+        It is critical for the internal data from the events to be returned,
+        upon `event_guts`.
+        input_info = {'input_kwarg': ('data_key', stream_number)}
+        Note that the stream number is assumed to be zero if not specified
+
+        output_info is designed to take the output tuple and map it back into
+        data_keys.
+        output_info = [('data_key', {'dtype': 'array', 'source': 'testing'})]
+        """
+        # TODO: this needs something super maybe a Base Class
+        Stream.__init__(self, child, children)
+        if output_info is None:
+            output_info = {}
+        if input_info is None:
+            input_info = {}
+        self.outbound_descriptor_uid = None
+        self.output_info = output_info
+        self.allowed_names = ['start', 'descriptor', 'event', 'stop']
+        self.i = None
+        self.run_start_uid = None
+        self.provenance = {}
+        self.event_failed = False
+
+        # If the stream number is not specified its zero
+        for k, v in input_info.items():
+            if len(v) < 2 or isinstance(v, str):
+                input_info[k] = (v, 0)
+            if isinstance(v[1], Stream):
+                input_info[k] = (v[0], self.children.index(v[1]))
+        self.input_info = input_info
+
+    def emit(self, x):
+        """ Push data into the stream at this point
+
+        This is typically done only at source Streams but can theoretically be
+        done at any point
+        """
+        if x is not None:
+            result = []
+            for parent in self.parents:
+                r = parent.update(x, who=self)
+                if type(r) is list:
+                    result.extend(r)
+                else:
+                    result.append(r)
+            return [element for element in result if element is not None]
+
+    def dispatch(self, nds):
+        name, docs = self.curate_streams(nds)
+        if name in self.allowed_names:
+            return getattr(self, name)(docs)
+
+    def update(self, x, who=None):
+        return self.emit(self.dispatch(x))
```

Comment:
But this could get annoying the second you try to go around this, eg tests.
I'm not certain what the best way to do this is. Adding it to the init seems like a good way but then you'd need to specify the db on every single stream function, which seems bad.